### PR TITLE
all code removed

### PR DESCRIPTION
**Pull request by skanes17**

### CODE REVIEW

File: src/components/buttons/AddRemoveButton.tsx [hunk 1]
```diff
@@ -1,36 +0,0 @@
-import React from "react";
```
Comment (devthedevel) [inline src/components/buttons/AddRemoveButton.tsx:1]: So many styling systems have the concept of 'Primary' or 'Secondary' or 'Neutral' or 'Accent' colors. Material UI and Daisy UI are two libraries off the top of my head that uses this paradigm. These colors are kind of related to 'actions': primary is a positive action, secondary could be a negative, etc. The semantics of the colors are up to you.

The idea is that you you don't directly use the color system; instead you use these named colors like `bg-primary`. The benefit of this is that you don't have to remember what action colors you choose for different components. If a `SubmitForm` has two buttons, one for submitting, and one for deleting then they can both utilize 'Primary' and 'Secondary' respectively.

It was earlier this week that I discovered you can do this with Tailwind if you want to. You can define your 'action colors' in the config, and then use them like `bg-primary` or `bg-neutral` wherever you want.

I'm writing this to bring a little context here. This component is called `AddRemoveButton` which takes a `buttonType` prop that styles accordingly. If the difference is just 'positive' and 'negative', then it might be worth considering setting up your own colors like above. You can even hook them into boolean props to use the components like `<Button primary>` or `<Button accent>`. 

You can also then add additional modifiers like `size` if you wanted to. The end result would be one generic `<Button>` component that can change its styling based on different props passed in. You could even build on top of this generic button to still create specific buttons if you wanted to.

None of this is a required change, just an alternative for thought.

File: src/components/buttons/AddRemoveButton.tsx [hunk 2]
```diff
@@ -1,36 +0,0 @@
...
-
-interface AddRemoveButtonProps {
-  children: string;
```
Comment (devthedevel) [inline src/components/buttons/AddRemoveButton.tsx:4]: I think if you want to make this more flexible (as in adding an icon on the button), you should change the children type to `React.JSXElement` I think.

File: src/components/buttons/AddRemoveButton.tsx (hunk 1 above)
Comment (devthedevel) [reply src/components/buttons/AddRemoveButton.tsx:1]: To add, if you had a lot of common styling being copied from button to button, you can then place that in the generic button component and then you only need to change it in one spot.

File: src/components/buttons/AutoButton.tsx [hunk 3]
```diff
@@ -1,17 +0,0 @@
-interface AutoButtonProps {
-  buttonText: "Auto";
```
Comment (devthedevel) [inline src/components/buttons/AutoButton.tsx:2]: I don't think you need `"Auto"` here: this can just be a string

File: src/components/buttons/Button.tsx [hunk 4]
```diff
@@ -1,32 +0,0 @@
-import React, { ReactNode } from "react";
-
-export interface ButtonProps {
```
Comment (devthedevel) [inline src/components/buttons/Button.tsx:3]: This looks similar to `AddRemoveButton`

File: src/components/buttons/AddRemoveButton.tsx [hunk 5]
```diff
@@ -1,36 +0,0 @@
...
-    case "remove":
-      bgColor = "bg-deepRed shadow-deepRed/50 hover:bg-red-600";
-      break;
-  }
-
-  return (
-    <button
-      type="button"
-      className={`shadow ${bgColor} mb-1 min-w-full self-end rounded py-1 text-xs font-semibold text-white`}
-      onClick={() => onClick()}
```
Comment (devthedevel) [inline src/components/buttons/AddRemoveButton.tsx:29]: Since `onClick` type does not contain any parameters (which suggests that you never want to pass any parameters), you can just shorten this to `onClick={onClick}`

File: src/components/cards/BldgCardImage.tsx [hunk 6]
```diff
@@ -1,45 +0,0 @@
-import React from "react";
-
-// images this component may need to use -- is there a better way?
-import archeryHut from "../../assets/images/archeryHut-01.jpg";
-import healingChamber from "../../assets/images/healingChamber-01.jpg";
-import mealHall from "../../assets/images/mealHall-01.jpg";
-import swordsmithy from "../../assets/images/swordsmithy-01.jpg";
-import townCenter from "../../assets/images/townCenter-01.jpg";
-import { BuildingType } from "../../types";
```
Comment (devthedevel) [inline src/components/cards/BldgCardImage.tsx:9]: So you could create an `images.ts` file that does all these exports and then re-exports them for you.

File: src/components/cards/BldgCardImage.tsx [hunk 7]
```diff
@@ -1,45 +0,0 @@
...
-    case "swordsmithy":
-      src = swordsmithy;
-      break;
-    case "townCenter":
-      src = townCenter;
-      break;
-    default:
-      src = "";
-      break;
-  }
```
Comment (devthedevel) [inline src/components/cards/BldgCardImage.tsx:38]: Not a fan of this to be honest. Instead of having to pass in a `buildingType` why not pass in an `imagePath` instead? That way you don't have to do switch statements to determine the image path, you don't have to mess with types either. You literally just pass in `../assets/images/mealhall.jpg` and you are done.

Better yet, use this with the above idea of re-exporting them through `images.ts`.

Then you can just do `src={props.imagePath}`.

File: src/components/cards/CardBgWithImage.tsx [hunk 8]
```diff
@@ -1,67 +0,0 @@
-import React, { ReactNode } from "react";
-
-interface CardBgWithImageProps {
-  cardStyle: "planning" | "combat";
-  lockedOrUnlockedUnits?: "locked" | "unlocked";
-  saturation: "oversaturated" | "normal" | "half" | "quarter" | "zero";
```
Comment (devthedevel) [inline src/components/cards/CardBgWithImage.tsx:6]: May be a good spot for enums

File: src/components/cards/CardBgWithImage.tsx [hunk 9]
```diff
@@ -1,67 +0,0 @@
...
-      break;
-    case "quarter":
-      saturate = "saturate-25";
-      break;
-    case "zero":
-      saturate = "saturate-0";
-      break;
-    default:
-      saturate = "saturate-100";
-  }
```
Comment (devthedevel) [inline src/components/cards/CardBgWithImage.tsx:47]: You could safe yourself one branch and do 

```
case "normal":
default:
  saturate: "saturate-100"
  break;
```

Switches allow you to _fallthrough_ branches. If you don't explicitly put `break` at the end of a branch, the code will continue to evaulate the rest of the branches and execute them if needed.

Also always use `break`

File: src/components/cards/CardFooter.tsx [hunk 10]
```diff
@@ -1,19 +0,0 @@
-interface CardFooterProps {
-  /* could add attack here */
-  tier: number;
-  health: number;
-}
-
-export default function CardFooter({ tier, health }: CardFooterProps) {
-  return (
-    <>
-      <div></div>
```
Comment (devthedevel) [inline src/components/cards/CardFooter.tsx:10]: This needed?

File: src/components/cards/CardHeader.tsx [hunk 11]
```diff
@@ -1,15 +0,0 @@
-import React from "react";
-
-interface NewCardHeaderProps {
-  cardName: string;
```
Comment (devthedevel) [inline src/components/cards/CardHeader.tsx:4]: You could argue that you use `children` here instead. If so then you would just do `<NewCardHeader>MyCardName</NewCardHeader>` instead of `<NewCardHeader cardName="">`. 

Not much of a difference really. If you plan on passing in icons or other html elements, then use children. If not, keep with the strings.

File: src/components/cards/CardHoverText.tsx [hunk 12]
```diff
@@ -1,30 +0,0 @@
-import React, { ReactNode } from "react";
-import { BaseResourceType, Resources } from "../../types";
-
-interface CardHoverTextProps {
-  lockedOrUnlockedUnits?: "locked" | "unlocked";
```
Comment (devthedevel) [inline src/components/cards/CardHoverText.tsx:5]: I noticed `"locked" | "unlocked"` has been used a lot....might be a good idea to break it out to its own type (or even enum)

File: src/components/cards/CardHoverText.tsx [hunk 13]
```diff
@@ -1,30 +0,0 @@
...
-  children: ReactNode;
-}
-
-export default function CardHoverText({
-  lockedOrUnlockedUnits,
-  children,
-}: CardHoverTextProps) {
-  let lockedAndUnlockedStyles;
-  switch (lockedOrUnlockedUnits) {
-    case "unlocked":
```
Comment (devthedevel) [inline src/components/cards/CardHoverText.tsx:15]: Again you can just use a fallthrough case here

File: src/components/cards/CardImageHeaderDescriptionContainer.tsx [hunk 14]
```diff
@@ -1,43 +0,0 @@
...
-        cardStyle="planning"
-        saturation={saturation}
-        bgImage={bgImage}
-      >
-        <NewCardHeader cardName={cardName} />
-        <CardHoverText>{children}</CardHoverText>
-      </CardBgWithImage>
-    </>
-  );
-}
```
Comment (devthedevel) [inline src/components/cards/CardImageHeaderDescriptionContainer.tsx:32]: Looks clean

File: src/components/cards/CardShowCount.tsx [hunk 15]
```diff
@@ -1,21 +0,0 @@
-import React from "react";
-
-interface CardShowCountProps {
-  countToShow: number | string;
```
Comment (devthedevel) [inline src/components/cards/CardShowCount.tsx:4]: Why `number | string`?

File: src/components/cards/ConstructBuilding.tsx [hunk 16]
```diff
@@ -1,123 +0,0 @@
...
-  CardHeader,
-  CardShowCount,
-  CardSymbol,
-  CardTemplate,
-} from "../cards";
-import CardImageHeaderDescriptionContainer from "./CardImageHeaderDescriptionContainer";
-
-interface ConstructBuildingProps {
-  buildings: Buildings;
-  setBuildings: Dispatch<SetStateAction<Buildings>>;
```
Comment (devthedevel) [inline src/components/cards/ConstructBuilding.tsx:26]: This assumes that you will always be passing some react `setState()` function here, which may not always be the case.

File: src/components/cards/ConstructBuilding.tsx [hunk 17]
```diff
@@ -1,123 +0,0 @@
...
-    }
-  };
-
-  const handleBuildClick = (buildingType: string) => {
-    // check that you've collected all required resources
-    const resourceCheck = resourceChecker(1, costsObject, resourcePool);
-
-    if (buildings[buildingType].underConstruction === false && resourceCheck) {
-      // set the building to be constructed
-      const clonedBuildingsData = cloneBasicObjectWithJSON(buildings);
```
Comment (devthedevel) [inline src/components/cards/ConstructBuilding.tsx:71]: Note to self, come back to this

File: src/components/cards/ConstructBuilding.tsx [hunk 18]
```diff
@@ -1,123 +0,0 @@
...
-  setResources: Dispatch<SetStateAction<Resources>>;
-  resourcePool: ResourcePool;
-  setResourcePool: Dispatch<SetStateAction<ResourcePool>>;
-  /* setResources: (resources: Resources) => void; */
-}
-
-export default function ConstructBuilding({
-  buildings,
-  setBuildings,
-  buildingType,
```
Comment (devthedevel) [inline src/components/cards/ConstructBuilding.tsx:40]: So one thing I noticed down below. You are always doing `buildings[buildingType]`. You can do one of two things here

1. `const building = buildings[buildingType]`. Do it once at the top and then just use it.
2. Don't pass in both `buildings` and `buildingType`. Just take a `building`. Let the parent component do `<ConstructBuilding building={buildings[buildingType]}` instead. Prefer this because theres no need for this component to know about all the rest of the buildings.

File: src/components/cards/ConstructBuilding.tsx [hunk 19]
```diff
@@ -1,123 +0,0 @@
...
-
-  const handleBuildClick = (buildingType: string) => {
-    // check that you've collected all required resources
-    const resourceCheck = resourceChecker(1, costsObject, resourcePool);
-
-    if (buildings[buildingType].underConstruction === false && resourceCheck) {
-      // set the building to be constructed
-      const clonedBuildingsData = cloneBasicObjectWithJSON(buildings);
-      clonedBuildingsData[buildingType].underConstruction = true;
-      setBuildings(clonedBuildingsData);
```
Comment (devthedevel) [inline src/components/cards/ConstructBuilding.tsx:73]: Ok I think I got a better idea of whats happening here. You want to to set a specific building's `underConstruction` to true and then update that in the `buildings` state.

So what you can do instead of cloning here, is pass the newly updated object back to the parent via a function. The function implementation in the parent will handle creating a new buildings state and only update the one particular building. It may seem like more work to add another function here, but you can be as specific as you want. You will also enjoy types if you set it up correctly. And you wouldnt be passing in `buildings` if you don't need it.

We may have to go over this in a little more detail next session.

File: src/components/cards/PreCombatCardTemplate.tsx [hunk 20]
```diff
@@ -1,61 +0,0 @@
-import React from "react";
-import { enemyColor, friendlyColor } from "../../gameData";
-import { ArmyColors, BaseUnit, Unit, UnitCounts, UnitType } from "../../types";
-
-interface PreCombatCardTemplateProps {
-  armyStyle: "friendly" | "enemy";
-  headerText: "Your Army" | "Enemy Army";
-  BASE_UNIT_DATA: BaseUnit;
```
Comment (devthedevel) [inline src/components/cards/PreCombatCardTemplate.tsx:8]: If this is only being used to determine the background image, then prefer not to pass in the entire `BaseUnit` but instead something like `bgImage: BaseUnit["bgImageSm"]`. More precise. Also YELLING IS NOT FUN

File: src/utils/produceScoutReport.tsx [hunk 21]
```diff
@@ -1,99 +0,0 @@
...
-) {
-  const enemyIsTwiceAsLarge = enemyArmy.length > 2 * friendlyArmy.length;
-  const enemyIsApproxSameSize =
-    Math.abs(enemyArmy.length - friendlyArmy.length) <
-    0.1 * friendlyArmy.length;
-  const enemyIsHalfSize = enemyArmy.length < 0.5 * friendlyArmy.length;
-  const enemyIsTwiceAsPowerful = enemyPowerLevel > friendlyPowerLevel * 2;
-  const enemyIsApproxSamePower =
-    Math.abs(enemyPowerLevel - friendlyPowerLevel) < 0.1 * friendlyPowerLevel;
-  const enemyIsHalfPowerful = enemyPowerLevel < 0.5 * friendlyPowerLevel;
```
Comment (devthedevel) [inline src/utils/produceScoutReport.tsx:18]: I think what you can do is just do `enemyArmy.length / friendlyArmy.length` to get a percentage. Then instead of doing the percent checking here you can do it in the parent code. This should allow you to add as many ranges as you like without having to explicitly do so here. Also since the ratio is just a number and not a list of possible booleans, its just more flexible.

Also don't return any UI (HTML elements) from utility code. They should be separated.

File: src/utils/resourceChecker.ts
Comment (devthedevel) [inline src/utils/resourceChecker.ts:1]: Mmmmmm not sure how I feel about this. I think `amount` should be a number and not `1 | 5`.

Is this function only returning a boolean? As in, if there is enough resources for 5 wood, then this returns true? If so this should be fine.

File: src/gameData/TutorialMessages.tsx
Comment (devthedevel) [inline src/gameData/TutorialMessages.tsx:1]: Again you shouldn't return UI stuff from non UI code

File: src/components/Game.tsx
Comment (devthedevel) [inline src/components/Game.tsx:1]: There is a lot going on in here, gonna come back to this over time

Comment (devthedevel) [reply src/components/Game.tsx:1]: I see that you are using devtools on every possible screen. It might be worth setting up the code so that the layout is a little more modular.

```
const currentScreen = // if planning then Planning, if combat then Combat, etc

return (
  {currentScreen}
  {devTools ? <DevTools> : undefined}
)
````

This will always keep devtools on the bottom of the page, regardless of whatever screen you are on.

Comment (devthedevel) [reply src/components/Game.tsx:1]: Also it might be worth hauling localstorage out into its own hook or perhaps just its own utility function. You can put some typing around it too.

File: src/utils/resourceChecker.ts
Comment (skanes17) [reply src/utils/resourceChecker.ts:1]: Changed to number

File: src/components/Game.tsx
Comment (skanes17) [reply src/components/Game.tsx:1]: How would I make my own hook for this?
